[docs][base] Show components & hooks API on the components page

mnajdova opened the pull request:
This PR converts the ~MUI Base~ Base UI's components pages to have tabs for the components & hooks apis. Changes done to make this possible:

- The `docs:api` cmd now generates the following pages:
  - /react-componentName/index.js - the initial page for the component (demos)
  - /react-componentName/[docsTab]/index.js - the page for the tabs APIs (components & hooks)
- The new pages use the `getLayout` method, that allows us to define the app frame once, so that the transition between the tabs (and pages for that matter) is smoother. The files are generated based on the md files and currently run only for the base docs, more specifically for the pages inside the `components/` folder. 
- The JSDocs of the components & hooks are also updated to point to the new API locations.
- the scripts are updated to not generate the old API pages, and they have been removed 

## Preview
Demos page: https://deploy-preview-35938--material-ui.netlify.app/base/react-tabs/
Components API: https://deploy-preview-35938--material-ui.netlify.app/base/react-tabs/components-api
Hooks API: https://deploy-preview-35938--material-ui.netlify.app/base/react-tabs/hooks-api


Todos:
- [x] add redirects for the old API pages

michaldudak commented:
Great initiative!
We probably could iterate a bit on the design, so the tabs are discoverable better.
It would also be great if we have links to individual tabs.

mnajdova commented:
> We probably could iterate a bit on the design, so the tabs are discoverable better.

I will wait on @danilo-leal 's input for this. For now, I moved them before the add and made them sticky so that they are visible when scrolling.

> It would also be great if we had links to individual tabs.

Done, luckily next.js has a [shallow routing](https://nextjs.org/docs/routing/shallow-routing) feature that we can use. Now the tabs are sharable via just copying the website address.

cherniavskii commented on docs/src/modules/components/ComponentPageTabs.js:
```diff
@@ -0,0 +1,38 @@
+  return (
+    <Box
+      className="component-tabs"
+      sx={{
+        position: 'sticky',
+        top: 5,
+        zIndex: 9999,
+      }}
+    >
+      <Tabs
+        value={activeTab}
+        onChange={(e, value) => setActiveTab(value)}
+        sx={{ fontSize: 'large' }}
```
To avoid sticky ad 🙂

```suggestion
      sx={{ display: 'inline' }}
    >
      <Tabs
        value={activeTab}
        onChange={(e, value) => setActiveTab(value)}
        sx={{
          fontSize: 'large',
          position: 'sticky',
          top: 5,
          zIndex: 9999,
        }}
```

mnajdova replied:
Thanks!

oliviertassinari commented:
Great to see us experiment 👍. I noticed a few bugs, but I would imagine that it doesn't matter for the goal of this PR. (The bugs are: the new tabs are visible over the drawer, the ad has a layout shift when loading, the page or the tab isn't server-side rendered & kept the same when reloading.)

Could we see a component that has more than **one** exported module to be migrated, so developers can experience what the UX would be like?

As a side thought, we likely have different UX options that optimize for different user journeys. I think that what will be key is to get the UX options that improve the most common journeys. I can think of the following ones:

1. I'm in my IDE, I click on the API page link in IntelliSense, I want to see the content in a very clear way. 
2. I'm in my IDE, I search on Google for the API of the module's name, I want to see the content in a very clear way.
3. I'm in my IDE, I switch to a tab that has MUI's docs open, I search for the module's name, I want to see the content in a very clear way.
4. I'm in my IDE, I switch to a tab that has MUI's docs open, I use the navigation links of the docs.
5. I'm on a page of the docs, I see a module that intrigue me (e.g. LoadingButton on the card page), I try to find more details. I might use one of the paths of 2, 3 or 4.
6. I'm on the component guide page, but I want to quickly access the API reference information. I might use one of the paths of 2, 3 or 4.

alexfauquette commented:
Nice to see it live 🎉

I'm wondering how the documentation of subcomponents API page will be handled. For example [`Table`](https://mui.com/material-ui/react-table/#api) which uses `TableContent`, `TableFooter`, ...

For the API page itself, I assume we can remove or disable the "demo" tab. But for the discoverability, I don't know how we can encourage to explore subcomponents documentation.

mnajdova commented:
I will update the PR to contain the changes for the `react-tabs` page, there are multiple components there.

mnajdova commented:
I've added an example of how we can show APIs for multiple components/hooks. You can see it here: https://deploy-preview-35938--material-ui.netlify.app/base/react-tabs/ I am simply iterating trough all components/hooks and showing all sections.

mnajdova commented on docs/src/modules/components/AppNavDrawerItem.js:
```diff
@@ -252,42 +253,44 @@ export default function AppNavDrawerItem(props) {
-      ) : (
-        children
-      )}
+      <NoSsr>
```
I don't like this, but I don't have a better idea at this moment. The API sidebar links now need to point to a link containing hash, as multiple hooks docs can be shown on the hooks api tab.

danilo-leal commented:
@mnajdova just pushed a few changes there! Had to go just a bit sideways ⎯ tweaking other components ⎯ to generally make the header section a bit more tuned in. Let me know what do you think about it!

mnajdova commented:
> @mnajdova just pushed a few changes there! Had to go just a bit sideways ⎯ tweaking other components ⎯ to generally make the header section a bit more tuned in. Let me know what do you think about it!

Looks good 👍

michaldudak commented:
<img width="716" alt="image" src="https://user-images.githubusercontent.com/4696105/218106437-8fbcd3dd-e03c-4029-ba79-6bc2e0fbd288.png">

Could we improve the spacing around headers a bit? In this case the "TabsListUnstyled API" should be more separated from the section above, as the bottom margin is larger than the top one

flaviendelangle commented:
![image](https://user-images.githubusercontent.com/3309670/218106847-d6568e06-0d44-49e9-ac14-76b77258afbf.png)

We have unwanted spaces in the import example

flaviendelangle reviewed (commented):
It looks great :partying_face: 

I'm wondering if we should keep one import example per component or have a single one at the top.

Right now we are duplicating the "You can learn about the difference by reading this guide on minimizing bundle size" on each component.

An alternative would be to do the following:

```tsx
import TabUnstyled from '@mui/base/TabUnstyled';
import TabsUnstyled from '@mui/base/TabsUnstyled';
import TabsListUnstyled from '@mui/base/TabsListUnstyled';
import TabPanelUnstyled from '@mui/base/TabPanelUnstyled';

// or
import { 
  TabUnstyled, 
  TabsUnstyled, 
  TabsListUnstyled, 
  TabPanelUnstyled 
}  from '@mui/base';
```

It makes the page shorter, but it makes the initial import paragraph longer.
For a set of tightly linked component like the tabs, I prefer to have one single import.
For a set of component with one major component and a set of minor dependencies (the `Menu` where most people don't use most of the `MenuXXX` components) the trade-off might not be the same.

michaldudak commented:
A minor design issue:

<img width="688" alt="image" src="https://user-images.githubusercontent.com/4696105/218108345-ee32f93f-3932-41cc-a8c2-2913a6e70715.png">

It would look better if the tabs panel was wider, so the blurred background doesn't end so abruptly and could match the blur from the header above.

danilo-leal commented:
> It would look better if the tabs panel was wider, so the blurred background doesn't end so abruptly and could match the blur from the header above.

Absolutely, that way, we could get somewhere closer to the design I've been working on:

<img width="600" alt="Screen Shot 2023-02-10 at 10 56 13" src="https://user-images.githubusercontent.com/67129314/218109359-eb971378-996e-4aa6-b465-4db4f0b616c1.png">

But it actually doesn't sound minor 😅 You tell me but I figure we'd have to do some slight restructuring on the docs layout components to get that!

⎯ 

- re **Spacing between component sections:** yup, totally. I'll work on that a bit too ⎯ was even curious to explore some divider action there.
- re **Duplicated content:** for sure, I can do these tweaks too to the docs.

oliviertassinari commented:
> how do we want to get feedback on the API tabs?

I think that it would be great to have an RFC about it. I also think that it would be great to have a 3rd hybrid option to play with: tabs with links to each individual API page. I think that the bulk of the problem today is that the API section **links are too** far for easy access, they are at the bottom of the guide component page, which is a lot to scroll or a lot of mouse distance to travel to if using the TOC => pain. So maybe if the links were in the tabs, then it would feel x2-x3 easier to access and hence fix the problem will keeping a very clear API reference section.

> put this in production

I can see the following UX problems on the PR preview:

1. Side nav active page scroll into view (all pages)
2. Anchor scroll top (only the new pages)
3. Active TOC detection (only the new pages)
4. Tabs server-side rendering (only the new pages)
5. Tabs background width (only the new pages)
6. Scroll reset between tabs (only the new pages)

Happy to provide more details if some are obscure.

danilo-leal commented:
Let me re-list the identified issues we need to tackle to understand whether this is going in the right direction as I've just tweaked some of them.

### General

- [x] **Anchor scroll top:** ensuring that when clicking at TOC elements, the title is clearly visible below the tabs.
- [ ] **Active TOC detection**: fine-tune just a bit the TOC active item detection now that the tabs add up on vertical scroll.
- [ ] **Tabs SSR**: ensure that content within tab panels is findable by Algolia and search engines.
- [x] Tabs width on mobile: ensure it spans the full width.

### Component API tab

- [ ] **Unnecessary import content on the API page**: since everything is now sitting on one page, we don't need to display the same info (+ code blocks) all over again. See [Flavien's comment](https://github.com/mui/material-ui/pull/35938#pullrequestreview-1293208135) for the proposed solution.
- [ ] **Wonky spacing between each component doc:** We may give a little bit more spacing between each component section, possibly exploring adding a divider there.

- - - - 

@oliviertassinari I'm wondering about your hypothesis. If we'd just throw links within the "Component API" tabs panel, for example, isn't an even shorter path to display the content there ⎯ as we're doing now?! I feel like this current design goes right at the core of the problem you've developed! Lastly, I don't fully understand items 1 & 7 of your list ⎯ would you mind expanding on them a bit?

mnajdova commented:
> <img width="716" alt="image" src="https://user-images.githubusercontent.com/4696105/218106437-8fbcd3dd-e03c-4029-ba79-6bc2e0fbd288.png">

@michaldudak nice catch, this was a regression from https://github.com/mui/material-ui/pull/35828, it is fixed in https://github.com/mui/material-ui/pull/36168

oliviertassinari commented:
>If we'd just throw links within the "Component API" tabs panel, for example, isn't an even shorter path to display the content there ⎯ as we're doing now?!

@danilo-leal I'm not sure, I feel like the tabs could work well.

> I feel like this current design goes right at the core of the problem you've developed!

An example the developer is on https://deploy-preview-35938--material-ui.netlify.app/base/react-tabs/#anatomy, he wants to find the `TabsUnstyled` prop available.

<img width="599" alt="Screenshot 2023-02-13 at 22 27 32" src="https://user-images.githubusercontent.com/3165635/218578440-dacf5c07-7989-4a6a-a56a-09f6e619826c.png">

His user journey with option 2 (where option 1 is HEAD):

1. Move his eye/cursor up to the tabs
2. Click on "Component API" in the header

<img width="311" alt="Screenshot 2023-02-13 at 22 24 35" src="https://user-images.githubusercontent.com/3165635/218577830-2630d3a7-449c-4dbb-8747-4cf8f22b116b.png">

3. Move his eye/cursor to the Table of contents, search for the component
4. Click on "TabsUnstyled"
5. Move back his attention to the body of the page

With option 3 (the hybrid approach), the distance to travel/information to process in step 3, 5 is reduced, and the rest stays the same.

> Lastly, I don't fully understand items 1 & 7 of your list ⎯ would you mind expanding on them a bit?

1. Is about https://deploy-preview-35938--material-ui.netlify.app/base/react-tabs/ that doesn't show the selected page of the sidenav in the viewport anymore.
7. Is about the user journey 7 of https://github.com/mui/material-ui/pull/35938#issuecomment-1404064718. I assume that we will redirect all the API reference URL pages to these new ones.

I have notice these others:

8. On https://deploy-preview-35938--material-ui.netlify.app/base/react-tabs/?docsTab=component-api#tabs-list-unstyled, there are more than one h1.
9. There is a blur discontinuation. Current:

<img width="835" alt="Screenshot 2023-02-13 at 22 32 27" src="https://user-images.githubusercontent.com/3165635/218579248-05d64808-da9a-4bc0-8c60-a44ef351c356.png">

Something in this order would feel better:

<img width="841" alt="Screenshot 2023-02-13 at 22 40 58" src="https://user-images.githubusercontent.com/3165635/218580889-829802a4-249a-4154-a1ac-81cbabc92929.png">

danilo-leal commented:
Gotcha ⎯ thanks for the detailed walkthrough of your thinking! I'm curious to hear what @mnajdova thinks about it, but I'm sensing that for the sake for simplicity, we could treat all Component API-related improvements ⎯ given they're now listing all exported modules ⎯ in a separated PR?! There are these problems y'all have listed (multiple H1s, spacing between each exported module, duplicated import content, etc.) that we could tackle there, isolatedly, _but_ with as a dependency to this PR (as they need to be fine-tuned for this whole tab design proposal to work well). What do you think?

⎯ ⎯ 

@oliviertassinari how do I get to this continued blur effect? Definitely much better!

oliviertassinari commented:
> how do I get to this continued blur effect? Definitely much better!

I copied what you did for mobile

mnajdova commented on docs/pages/_app.js:
```diff
@@ -200,10 +200,11 @@ AppWrapper.propTypes = {
   return (
     <AppWrapper emotionCache={emotionCache} pageProps={pageProps}>
-      <Component {...pageProps} />
+      {getLayout(<Component {...pageProps} />)}
```
Needed to support the `Page.getLayout` defined in the new pages.

mnajdova commented on docs/src/modules/components/AdGuest.js:
```diff
@@ -3,7 +3,7 @@ import PropTypes from 'prop-types';
 import { AdContext } from 'docs/src/modules/components/AdManager';
 
-export default function AdGuest(props) {
+export default function AdGuest({ classSelector = '.description', ...props }) {
```
The `classSelector` prop allows other elements to be connected to the ad, for example the tabs element, instead of the description. By default (without the prop being provided), the ad will be placed below the description - same as before.

mnajdova commented on docs/src/modules/components/AppLayoutDocs.js:
```diff
@@ -37,20 +37,30 @@ const Main = styled('main', {
-  shouldForwardProp: (prop) => prop !== 'disableAd',
-})(({ disableAd, theme }) => {
+  shouldForwardProp: (prop) => prop !== 'disableAd' && prop !== 'hasTabs',
+})(({ disableAd, hasTabs, theme }) => {
```
Needed in order to adjust a bit the styles when tabs are visible on the page.

mnajdova commented on docs/src/modules/components/AppTableOfContents.js:
```diff
@@ -124,6 +124,7 @@ const shouldShowJobAd = () => {
   return true;
 };
 
+// TODO: Add custom table of contents for the pages containing multiple tabs
```
```suggestion
```

Done

mnajdova commented on packages/api-docs-builder/buildApiUtils.ts:
```diff
@@ -520,3 +543,198 @@ export const stringifySymbol = (symbol: ts.Symbol, project: TypeScriptProject) =
   return formatType(rawType);
 };
+
+export function updateComponentPages() {
```
This is the method that generates the new pages (`react-component/index.js` and `react-component/[docsTab]/index.js`). Current this is configured only for MUI Base.

mnajdova commented on docs/src/modules/components/Link.tsx:
```diff
@@ -88,8 +88,12 @@ const Link = React.forwardRef<HTMLAnchorElement, LinkProps>(function Link(props,
 
   const router = useRouter();
   const pathname = typeof href === 'string' ? href : href?.pathname;
+  const routerPathname = router.pathname.replace('/[docsTab]', '');
```
The active class name shouldn't depend on which tab is active.

mnajdova commented on docs/src/modules/components/MarkdownDocsV2.js:
```diff
@@ -0,0 +1,256 @@
+  mode: PropTypes.oneOf(['light', 'dark']),
+};
+
+export default function MarkdownDocs(props) {
```
Once we migrate all docs to use tabs, the `MarkdownDocs.js` will be replaced by `MarkdownDocsV2.js`. I isolated the two, so that we won't accidently break something in the other pages.

mnajdova commented on packages/api-docs-builder/buildApiUtils.ts:
```diff
@@ -520,3 +564,172 @@ export const stringifySymbol = (symbol: ts.Symbol, project: TypeScriptProject) =
+      const tokens = markdown.pathname.split('/');
+      const name = tokens[tokens.length - 1];
+      const importStatement = `docs/data${markdown.pathname}/${name}.md`;
+      const demosSource = `
```
This page doesn't necessarily need to be generated, it can be copied as previously when adding the demos page, but I added this step to ensure that the `getLayout` method would be used, as it is not there on all pages at this moment. Once we migrate all pages to the new structure, we can remove this step.

alexfauquette reviewed (commented):
I did not find why but the ad loading has a strange behavior between API and demo pages.

If you switch between the two API pages, no reload of the Ad. But with demos it re-loads 🤔

[Screencast from 17-03-2023 14:58:21.webm](https://user-images.githubusercontent.com/45398769/225926048-aeeafa33-2ef1-439b-8626-3777b07c3c24.webm)

mnajdova commented:
> I did not find why but the ad loading has a strange behavior between API and demo pages.
> 
> If you switch between the two API pages, no reload of the Ad. But with demos it re-loads 🤔

Hmm, the only difference between the two is that the demos page is a different page: `/react-button/index.js`, while the component & hook tabs both map to `/react-button/[docsTab]/index.js`

siriwatknp commented:
Some improvements I see:
- The alignment of the icon

  <img width="465" alt="image" src="https://user-images.githubusercontent.com/18292247/227152455-b5ad297d-6851-4dfe-a8c5-f95fd7e1900c.png">

- The rerendering of the ads is pretty annoying. Not sure if it is possible to keep the same one between those pages.
- I think the background of the tabs should be opaque white without blur filter.

  <img width="866" alt="image" src="https://user-images.githubusercontent.com/18292247/227153619-dc8ca563-f0b0-4c52-a8c6-bd76907bb5e8.png">

siriwatknp commented on packages/api-docs-builder/buildApi.ts:
```diff
@@ -277,14 +276,54 @@ async function run(argv: yargs.ArgumentsCamelCase<CommandOptions>) {
+        // @ts-ignore
+        const { value } = build;
+        const { name, demos } = value;
+        const lastIdx = demos.length > 0 ? demos[0].demoPathname.indexOf('#') : -1;
+
+        let pathname = null;
+
+        if (demos.length > 0) {
+          pathname =
+            lastIdx >= 0
+              ? demos[0].demoPathname.substr(0, demos[0].demoPathname.indexOf('#'))
+              : demos[0].demoPathname;
+        }
+
+        if (pathname !== null) {
+          apiLinks.push({
+            pathname: `${pathname}${name.startsWith('use') ? 'hooks-api' : 'components-api'}`,
+            hash: kebabCase(name),
+            title: name,
+          });
+        }
```
A comment on this step would be nice.

mnajdova replied:
done

siriwatknp commented on packages/api-docs-builder/buildApiUtils.ts:
```diff
@@ -520,3 +564,137 @@ export const stringifySymbol = (symbol: ts.Symbol, project: TypeScriptProject) =
   return formatType(rawType);
 };
+
+export function generateApiPages() {
```
Does it make sense to name this function specific to Base? like `generateBaseUIApiPages`?

mnajdova replied:
For now yes, let's make it explicit. In the future if we add this to all project, we can make it part of the project's definition.

danilo-leal reviewed (commented):
Woo, this is super exciting ⎯ awesome work all around, we're almost at the finish line with this first iteration! Here are some of my thoughts so far, and none of them are really visual:

## 1️⃣ Scrolling position

Should we save it? I have a feeling that might be frustrating for the switching of tabs to always reset the scroll to the top. I imagine the use case, as briefly demonstrated in the video:

https://user-images.githubusercontent.com/67129314/227645870-6192c0fc-52e5-41cb-b86a-4be6be25ba0a.mov

<br />

1. I'm seeing something very specific in the "Slot props" section, for example.
2. I want to double-check something related to it on the Component API, so I go there.
3. Going back to the Demos tab, I'm at the top again.
4. I need now to either scroll down or use the TOC.

If that's something that happens frequently, it'll probably get annoying...

## 2️⃣ Repeated content

Should we optimize the way we display content that is now repeated given the API/hook pages were unified? I'm particularly thinking about these pieces:

- _"You can learn about the difference by reading this guide on minimizing bundle size"_.
- _"Props of the native component are also available."_
- _"The ref is forwarded to the root element._

A way that I imagined we could do this is by adding a "Disclaimers" section at the top. In hindsight, I'm not 100% sure if that's elegant enough 😅 Definitely open to iterating, but something in this direction might be good!

<img width="600" alt="Screen Shot 2023-03-24 at 18 32 23" src="https://user-images.githubusercontent.com/67129314/227645900-ea00bc3d-1fdd-4efa-85e8-ab3a178a0bb7.png">

## 3️⃣ Dividers between components/hooks

Each components/hook block is too glued to one another, vertically, I think. Here's what I'm talking about:

<img width="600" alt="Screen Shot 2023-03-24 at 18 23 30" src="https://user-images.githubusercontent.com/67129314/227645896-605eeb1f-d144-4230-98c8-6483f3c8cd4a.png">

It might be good to add a horizontal divider there to space them out a bit, making it easier to scan. Here's how I pictured this could be:

<img width="600" alt="Screen Shot 2023-03-24 at 18 32 34" src="https://user-images.githubusercontent.com/67129314/227645905-6615e767-7cf2-4a22-97d5-0fc31d64b5a6.png">

mnajdova commented:
> The alignment of the icon

Fixes, the main problem was that I used the icon button on h1 element, which was incorrect. Fixing the levels of the headings fixed this issue too.

> The rerendering of the ads is pretty annoying. Not sure if it is possible to keep the same one between those pages.

Not sure how if this be configured.

> I think the background of the tabs should be opaque white without blur filter.

Changed, looks much better. I've removed the alpha in dark mode too.

> Scrolling position

@danilo-leal we discussed this in some of the previous threads and decided to remove the scroll, mainly because the scroll from a previous tab can lead you to a random place on the new tab, which yield worse experience.

> repeated content

The only repeated content we can abstract is the one related for the import (how it affects bundle size). Everything else is dynamically extracted from the component, so it can be different between different components. As this is the only thing that is common, I would probably just leave it as is, mainly because I expect people to navigate directly to a component's API, so it would be nice to have all content without needing to scroll on the top to see this additional information.

> dividers between components/hooks

This is why I ended up with: 
![image](https://user-images.githubusercontent.com/4512430/228189540-401764b0-3d22-499d-89dd-2d0101eee30e.png)
![image](https://user-images.githubusercontent.com/4512430/228189662-981ff72d-a5bd-4d6b-9054-ac3e46bf7780.png)

danilo-leal reviewed (approved):
Looks great to me ⎯ excellent work 🎉

flaviendelangle commented:
![image](https://user-images.githubusercontent.com/3309670/228201045-cdfdc747-ede4-4801-b392-7362f798615a.png)

Should we remove the `Default` column when it's fully empty ?

mnajdova commented:
> Should we remove the Default column when it's fully empty?

That makes sense 👍, I will create a separate PR for this, as it would affect other API pages too, just to make sure I won't create regressions.

flaviendelangle reviewed (approved):
Looks great :birthday:

oliviertassinari commented on docs/pages/base/react-badge/[docsTab]/index.js:
```diff
@@ -0,0 +1,48 @@
+  };
+};
+
+export const getStaticProps = () => {
```
In the past, we struggled to make the translation work with this approach: https://github.com/vercel/next.js/issues/15315#issuecomment-735275969. I'm curious to see which translation approach React takes in https://github.com/reactjs/react.dev/issues/4135.

oliviertassinari commented on docs/src/modules/components/AppLayoutDocs.js:
```diff
@@ -37,20 +37,30 @@ const Main = styled('main', {
+          marginBottom: 40,
+        },
+      }),
+      ...(hasTabs && {
+        '&& .component-tabs .MuiTabs-root': {
+          marginBottom: 193,
+        },
+        '&& .component-tabs.ad .MuiTabs-root': {
+          marginBottom: 35,
+        },
+      }),
```
There is a layout shift when the ad loads:

<img width="769" alt="Frame 1 (1)" src="https://user-images.githubusercontent.com/3165635/229188573-d5e45c53-917c-41e0-9036-5fb796fc7d4f.png">

oliviertassinari commented on docs/src/modules/components/MarkdownDocsV2.js:
```diff
@@ -0,0 +1,256 @@
+      disableToc={disableToc}
+      location={location}
+      title={title}
+      toc={activeToc}
```
https://deploy-preview-35938--material-ui.netlify.app/base/react-tabs/components-api/#tabs-unstyled highlights the wrong header:

<img width="155" alt="Screenshot 2023-03-31 at 18 56 43" src="https://user-images.githubusercontent.com/3165635/229183096-f98f196f-2cfa-4873-9ce5-001f2b914ba9.png">

oliviertassinari commented on docs/data/base/pagesApi.js:
```diff
@@ -1,42 +1,169 @@
+    title: 'useTabPanel',
+  },
+  { pathname: '/base/react-tabs/hooks-api', hash: 'use-tabs', title: 'useTabs' },
+  {
+    pathname: '/base/react-tabs/hooks-api',
+    hash: 'use-tabs-list',
+    title: 'useTabsList',
+  },
```
Something is off, we can see it 4 times

<img width="290" alt="Screenshot 2023-04-01 at 15 06 31" src="https://user-images.githubusercontent.com/3165635/229290855-43f34857-2b95-4476-8294-51528fa94955.png">

https://deploy-preview-35938--material-ui.netlify.app/base/react-tabs/hooks-api/#use-tab-panel

mnajdova replied:
Fixed in https://github.com/mui/material-ui/pull/36764

oliviertassinari reviewed (commented):
Alright, we are getting there :)!

I did a deep dive into the changes with the thing that I could find, in addition to the inline comments:

1. #36743 fixed double redirections
2. #36732 is a fun one I took care of, this isn't directly related to these changes.

---

I think that it could be interesting to explore how we can have generic tab support in the docs infra, like https://stripe.com/docs/quotes/create?testing-method=without-code. Right now, it feels like it's not something MUI X or Toolpad could use.

---

In terms of UX, now that I can fully play with it, I'm personally in the camp of one page per exported module to consume the content is better/simpler to use. As we are making each of these pages more comprehensive (= unreadable toc, and = more scroll to), I think that this reinforces this feeling. However, the new UI to access the reference is better.

I think that as a next step, it would be great to move https://deploy-preview-35938--material-ui.netlify.app/material-ui/react-button/#api to the top of the page so it can be easily accessed (e.g. https://vuetifyjs.com/en/components/expansion-panels/#api) and to wait to see how developers are enjoying the different UX. 

In the meantime, I think that it would make more impact to invest time in fixing #34085 considering the volume of requests we have about it, and that it's not a too big of a task.

mnajdova commented on docs/data/base/pagesApi.js:
```diff
@@ -1,42 +1,169 @@
+  },
+  {
+    pathname: '/base/react-snackbar/hooks-api',
+    hash: 'use-snackbar',
+    title: 'useSnackbar',
```
Changed in https://github.com/mui/material-ui/pull/36764

oliviertassinari commented:
https://app.ahrefs.com/site-audit/3524616/59/data-explorer?columns=pageRating%2Curl%2Ctraffic%2ChttpCode%2Cdepth%2Ccompliant%2CincomingAllLinks%2Corigin&filterId=91013c8aafad4fe1f7a0ad504eccad42&issueId=c64da643-d0f4-11e7-8ed1-001e67ed4656&sorting=-pageRating surface an intresting issue:

1. Open https://mui.com/material-ui/react-tabs/#api
2. It links to https://mui.com/base/api/tabs-list-unstyled/
3. Click on it, it's a 404

I think that we should fix the redirection that lead to a 404, and then, update the link in the markdown so there is no redirections.

mnajdova commented on docs/src/modules/components/MarkdownDocsV2.js:
```diff
@@ -0,0 +1,256 @@
+          </Wrapper>
+        )}
+        {commonElements}
+        <Box {...(activeTab !== '' && { sx: { display: 'none' }, 'aria-hidden': true })}>
+          {rendered.slice(i, rendered.length - 1).map((renderedMarkdownOrDemo, index) => (
+            <MarkdownElement
+              key={`demos-section-${index}`}
+              renderedMarkdownOrDemo={renderedMarkdownOrDemo}
+              WrapperComponent={Wrapper}
+              wrapperProps={wrapperProps}
+              srcComponents={srcComponents}
```
Oops a left over from the initial implementation of the page, where we didn't export different pages for the tabs.
